pkg/nfd-topology-updater/nfd-topology-updater.go

```diff
@@ -133,19 +136,18 @@ func (w *nfdTopologyUpdater) Run() error {
 
 	klog.V(2).Infof("resAggr is: %v\n", resAggr)
 
-	crTrigger := time.NewTicker(w.resourcemonitorArgs.SleepInterval)
 	for {
 		select {
-		case <-crTrigger.C:
-			klog.Infof("Scanning")
+		case info := <-w.eventSource:
```

Comment:
We should create and configure the fs watcher, then filter the events we received from it and passing only the relevant one to the topology-updater select loop.

There is quite some work to do, hence the reason for moving this logic into  a separate package

Reply:
I was thinking something like this:
```
+       var crTrigger *time.Ticker
+       if w.resourcemonitorArgs.SleepInterval > 0 {
+               crTrigger = time.NewTicker(w.resourcemonitorArgs.SleepInterval)
+       }
+       ch, err := createFSWatcherEvent([]string{kubeletDirPath})
+       if err != nil {
+               return fmt.Errorf("failed to obtain node resource information: %w", err)
+       }
        for {
                select {
                case <-crTrigger.C:
-                       klog.Infof("Scanning")
-                       podResources, err := resScan.Scan()
-                       utils.KlogDump(1, "podResources are", "  ", podResources)
-                       if err != nil {
-                               klog.Warningf("Scan failed: %v", err)
-                               continue
-                       }
-                       zones = resAggr.Aggregate(podResources)
-                       utils.KlogDump(1, "After aggregating resources identified zones are", "  ", zones)
-                       if !w.args.NoPublish {
-                               if err = w.updateNodeResourceTopology(zones); err != nil {
-                                       return err
-                               }
-                       }
-
+                       w.runUpdater(resScan, resAggr)
                        if w.args.Oneshot {
                                return nil
                        }

+               case e := <-n.fsEvent:
+                       klog.V(5).Infof("fsnotify event from file %q: %q received", e.Name, e.Op)
+                       if e.Name == CPUStateFile ||
+                               e.Name == MemoryStateFile ||
+                               e.Name == DevicesStateFile {
+                               i := Info{Event: FSUpdate}
+                               w.runUpdater(resScan, resAggr)
+                       }
                case <-w.stop:
                        klog.Infof("shutting down nfd-topology-updater")
                        return nil
@@ -164,6 +164,28 @@ func (w *nfdTopologyUpdater) Run() error {

 }

+func (w *nfdTopologyUpdater) runUpdater(resScan resourcemonitor.ResourcesScanner, resAggr resourcemonitor.ResourcesAggregator) {
+       zones := w.aggregateZones(resScan, resAggr)
+       utils.KlogDump(1, "After aggregating resources identified zones are", "  ", zones)
+       if !w.args.NoPublish {
+               if err := w.updateNodeResourceTopology(zones); err != nil {
+                       klog.Warningf("cannot update NodeResourceTopology: %s", err.Error())
+               }
+       }
+}
+
+func (w *nfdTopologyUpdater) aggregateZones(resScan resourcemonitor.ResourcesScanner, resAggr resourcemonitor.ResourcesAggregator) v1alpha1.ZoneList {
+       klog.Infof("Scanning")
+       podResources, err := resScan.Scan()
+       utils.KlogDump(1, "podResources are", "  ", podResources)
+       if err != nil {
+               klog.Warningf("Scan failed: %v", err)
+               return nil
+       }
+
+       return resAggr.Aggregate(podResources)
+}
+
```
we could just put common logic to new funcs and remove necessity of having separate struct.

Reply:
I tend to agree that inline everything might be clearer, but IMVHO decoupling the event source generator from the main loop has greater benefits: 
For example you can generate your own source of events if needed, without changing the main loop logic.
This is useful especially for testing, in case you want to mock notification events. 
@PiotrProkop WDYT?

Reply:
This makes sense. I have no strong objections against putting events sources behind this abstraction, just wanted to indicate that this could be simplified. Both approaches works for me 😄